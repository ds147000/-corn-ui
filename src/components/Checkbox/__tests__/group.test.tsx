import React from 'react'
import { fireEvent, render, waitFor } from '@testing-library/react'
import CheckBoxGroup from '../group'
import Checck from '../checkbox'

test('基本使用', () => {
  const screen = render(
    <CheckBoxGroup>
      <Checck value="1">选项1</Checck>
      <Checck value="2">选项2</Checck>
    </CheckBoxGroup>
  )

  screen.getByText('选项1')
  screen.getByText('选项2')

  expect(screen.container).toMatchSnapshot()
})

test('选中', async () => {
  const onChange = jest.fn()
  const screen = render(
    <CheckBoxGroup onChange={onChange} >
      <Checck value="1">选项1</Checck>
      <Checck value="2">选项2</Checck>
    </CheckBoxGroup>
  )

  fireEvent.click(screen.getByText('选项1'))
  fireEvent.click(screen.getByText('选项2'))

  await waitFor(() => expect(screen.container).toMatchSnapshot())
  await waitFor(() => expect(onChange).toHaveBeenCalledTimes(2))
  await waitFor(() => expect(onChange.mock.calls[0][0]).toEqual(['1']))
  await waitFor(() => expect(onChange.mock.calls[1][0]).toEqual(['1', '2']))
})

test('取消选中', async () => {
  const onChange = jest.fn()
  const screen = render(
    <CheckBoxGroup onChange={onChange} >
      <Checck value="1">选项1</Checck>
      <Checck value="2">选项2</Checck>
    </CheckBoxGroup>
  )

  fireEvent.click(screen.getByText('选项1'))
  fireEvent.click(screen.getByText('选项2'))
  fireEvent.click(screen.getByText('选项1'))
  fireEvent.click(screen.getByText('选项2'))

  await waitFor(() => expect(screen.container).toMatchSnapshot())
  expect(onChange).toHaveBeenCalledTimes(4)
  expect(onChange.mock.calls[0][0]).toEqual(['1'])
  expect(onChange.mock.calls[1][0]).toEqual(['1', '2'])
  expect(onChange.mock.calls[2][0]).toEqual(['2'])
  expect(onChange.mock.calls[3][0]).toEqual([])
})

test('取消选中 和默认值', async () => {
  const onChange = jest.fn()
  const screen = render(
    <CheckBoxGroup onChange={onChange} defaultValue={['1', '2']} >
      <Checck value="1">选项1</Checck>
      <Checck value="2">选项2</Checck>
    </CheckBoxGroup>
  )

  fireEvent.click(screen.getByText('选项1'))
  fireEvent.click(screen.getByText('选项2'))

  await waitFor(() => expect(screen.container).toMatchSnapshot())
  await waitFor(() => expect(onChange).toHaveBeenCalledTimes(2))
  await waitFor(() => expect(onChange.mock.calls[0][0]).toEqual(['2']))
  await waitFor(() => expect(onChange.mock.calls[1][0]).toEqual([]))
})

test('获取值', () => {
  const ref = React.createRef<CheckBoxGroup>()
  render(
    <CheckBoxGroup ref={ref}>
      <Checck value="1">选项1</Checck>
      <Checck value="2">选项2</Checck>
    </CheckBoxGroup>
  )

  expect(ref.current.getValue()).toEqual([])
})


test('获取值 和 默认值', () => {
  const ref = React.createRef<CheckBoxGroup>()
  render(
    <CheckBoxGroup ref={ref} defaultValue={['1', '2']}>
      <Checck value="1">选项1</Checck>
      <Checck value="2">选项2</Checck>
    </CheckBoxGroup>
  )

  expect(ref.current.getValue()).toEqual(['1', '2'])
})

test('操作后获取值', async () => {
  const ref = React.createRef<CheckBoxGroup>()
  const screen = render(
    <CheckBoxGroup ref={ref}>
      <Checck value="1">选项1</Checck>
      <Checck value="2">选项2</Checck>
    </CheckBoxGroup>
  )

  fireEvent.click(screen.getByText('选项1'))

  await waitFor(() => expect(ref.current.getValue()).toEqual(['1']))
})

test('操作后获取值2', async () => {
  const ref = React.createRef<CheckBoxGroup>()
  const screen = render(
    <CheckBoxGroup ref={ref}>
      <Checck value="1">选项1</Checck>
      <Checck value="2">选项2</Checck>
    </CheckBoxGroup>
  )

  fireEvent.click(screen.getByText('选项1'))
  fireEvent.click(screen.getByText('选项2'))

  await waitFor(() => expect(ref.current.getValue()).toEqual(['1', '2']))
})

test('操作后获取值 重置', async () => {
  const ref = React.createRef<CheckBoxGroup>()
  const screen = render(
    <CheckBoxGroup ref={ref}>
      <Checck value="1">选项1</Checck>
      <Checck value="2">选项2</Checck>
    </CheckBoxGroup>
  )

  fireEvent.click(screen.getByText('选项1'))
  fireEvent.click(screen.getByText('选项2'))
  ref.current.reset()
  await waitFor(() => expect(ref.current.getValue()).toEqual([]))
})

test('单选模式', async () => {
  const onChange = jest.fn()
  const screen = render(
    <CheckBoxGroup onChange={onChange} radio >
      <Checck value="1">选项1</Checck>
      <Checck value="2">选项2</Checck>
      <Checck value="3">选项3</Checck>
      <Checck value="4">选项4</Checck>
    </CheckBoxGroup>
  )

  fireEvent.click(screen.getByText('选项1'))
  fireEvent.click(screen.getByText('选项2'))
  fireEvent.click(screen.getByText('选项3'))
  fireEvent.click(screen.getByText('选项4'))

  await waitFor(() => expect(screen.container).toMatchSnapshot())
  await waitFor(() => expect(onChange.mock.calls[3][0]).toEqual(['4']))
})

test('设置值方法', async () => {
  const ref = React.createRef<CheckBoxGroup>()
  render(
    <CheckBoxGroup ref={ref} >
      <Checck value="1">选项1</Checck>
      <Checck value="2">选项2</Checck>
    </CheckBoxGroup>
  )

  ref.current.setValue(['1'])

  await waitFor(() => expect(ref.current.getValue()).toEqual(['1']))
})

test('全选', async () => {
  const ref = React.createRef<CheckBoxGroup>()
  render(
    <CheckBoxGroup ref={ref} >
      <Checck value="1">选项1</Checck>
      <Checck value="2">选项2</Checck>
      <Checck value="3">选项3</Checck>
    </CheckBoxGroup>
  )

  ref.current.selectAll()

  await waitFor(() => expect(ref.current.getValue()).toEqual(['1', '2', '3']))
})

test('全选', async () => {
  const ref = React.createRef<CheckBoxGroup>()
  render(
    <CheckBoxGroup ref={ref} >
      <Checck value="1">选项1</Checck>
      <Checck value="2">选项2</Checck>
      <Checck value="3">选项3</Checck>
      <Checck value="3">选项3</Checck>
    </CheckBoxGroup>
  )

  ref.current.selectAll()

  await waitFor(() => expect(ref.current.getValue()).toEqual(['1', '2', '3']))
})

test('单选 全选', async () => {
  const ref = React.createRef<CheckBoxGroup>()
  render(
    <CheckBoxGroup ref={ref} radio >
      <Checck value="1">选项1</Checck>
      <Checck value="2">选项2</Checck>
      <Checck value="3">选项3</Checck>
    </CheckBoxGroup>
  )

  ref.current.selectAll()

  await waitFor(() => expect(ref.current.getValue()).toEqual([]))
})

test('单选选择', async () => {
  const ref = React.createRef<CheckBoxGroup>()
  const screen = render(
    <CheckBoxGroup ref={ref} radio >
      <Checck value="1">选项1</Checck>
      <Checck value="2">选项2</Checck>
      <Checck value="3">选项3</Checck>
    </CheckBoxGroup>
  )

  fireEvent.click(screen.getByText('选项2'))
  fireEvent.click(screen.getByText('选项2'))

  await waitFor(() => expect(ref.current.getValue()).toEqual(['2']))
})
