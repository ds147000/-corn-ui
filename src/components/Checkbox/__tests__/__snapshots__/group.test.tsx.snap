// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`单选模式 1`] = `
<div>
  <div
    class=""
  >
    <input
      class="corn-checkbox-group-hide"
      data-testid="check"
      readonly=""
      value="\\"4\\""
    />
    <div
      class="corn-if corn-ac corn-checkbox"
      data-testid="corn-checkbox"
    >
      <input
        class="corn-checkbox-hide"
        data-testid="check"
        name=""
        readonly=""
        value="1"
      />
      <div
        class="taro-img corn-img corn-checkbox-icon"
      >
        <img
          class="taro-img__mode-scaletofill"
          src="https://assets.xrkmm.cn/u/2000001252935290/dc949eed-eb8c-4ea5-bea4-c45a870cf0e7.png"
        />
      </div>
      选项1
    </div>
    <div
      class="corn-if corn-ac corn-checkbox"
      data-testid="corn-checkbox"
    >
      <input
        class="corn-checkbox-hide"
        data-testid="check"
        name=""
        readonly=""
        value="2"
      />
      <div
        class="taro-img corn-img corn-checkbox-icon"
      >
        <img
          class="taro-img__mode-scaletofill"
          src="https://assets.xrkmm.cn/u/2000001252935290/dc949eed-eb8c-4ea5-bea4-c45a870cf0e7.png"
        />
      </div>
      选项2
    </div>
    <div
      class="corn-if corn-ac corn-checkbox"
      data-testid="corn-checkbox"
    >
      <input
        class="corn-checkbox-hide"
        data-testid="check"
        name=""
        readonly=""
        value="3"
      />
      <div
        class="taro-img corn-img corn-checkbox-icon"
      >
        <img
          class="taro-img__mode-scaletofill"
          src="https://assets.xrkmm.cn/u/2000001252935290/dc949eed-eb8c-4ea5-bea4-c45a870cf0e7.png"
        />
      </div>
      选项3
    </div>
    <div
      class="corn-if corn-ac corn-checkbox"
      data-testid="corn-checkbox"
    >
      <input
        class="corn-checkbox-hide"
        data-testid="check"
        name=""
        readonly=""
        value="4"
      />
      <div
        class="taro-img corn-img corn-checkbox-icon"
      >
        <img
          class="taro-img__mode-scaletofill"
          src="https://assets.xrkmm.cn/u/2000001252935290/97f184d7-4fda-43f1-bd56-8313ee8a2b94.png"
        />
      </div>
      选项4
    </div>
  </div>
</div>
`;

exports[`取消选中 1`] = `
<div>
  <div
    class=""
  >
    <input
      class="corn-checkbox-group-hide"
      data-testid="check"
      readonly=""
      value="[]"
    />
    <div
      class="corn-if corn-ac corn-checkbox"
      data-testid="corn-checkbox"
    >
      <input
        class="corn-checkbox-hide"
        data-testid="check"
        name=""
        readonly=""
        value="1"
      />
      <div
        class="taro-img corn-img corn-checkbox-icon"
      >
        <img
          class="taro-img__mode-scaletofill"
          src="https://assets.xrkmm.cn/u/2000001252935290/dc949eed-eb8c-4ea5-bea4-c45a870cf0e7.png"
        />
      </div>
      选项1
    </div>
    <div
      class="corn-if corn-ac corn-checkbox"
      data-testid="corn-checkbox"
    >
      <input
        class="corn-checkbox-hide"
        data-testid="check"
        name=""
        readonly=""
        value="2"
      />
      <div
        class="taro-img corn-img corn-checkbox-icon"
      >
        <img
          class="taro-img__mode-scaletofill"
          src="https://assets.xrkmm.cn/u/2000001252935290/dc949eed-eb8c-4ea5-bea4-c45a870cf0e7.png"
        />
      </div>
      选项2
    </div>
  </div>
</div>
`;

exports[`取消选中 和默认值 1`] = `
<div>
  <div
    class=""
  >
    <input
      class="corn-checkbox-group-hide"
      data-testid="check"
      readonly=""
      value="[]"
    />
    <div
      class="corn-if corn-ac corn-checkbox"
      data-testid="corn-checkbox"
    >
      <input
        class="corn-checkbox-hide"
        data-testid="check"
        name=""
        readonly=""
        value="1"
      />
      <div
        class="taro-img corn-img corn-checkbox-icon"
      >
        <img
          class="taro-img__mode-scaletofill"
          src="https://assets.xrkmm.cn/u/2000001252935290/dc949eed-eb8c-4ea5-bea4-c45a870cf0e7.png"
        />
      </div>
      选项1
    </div>
    <div
      class="corn-if corn-ac corn-checkbox"
      data-testid="corn-checkbox"
    >
      <input
        class="corn-checkbox-hide"
        data-testid="check"
        name=""
        readonly=""
        value="2"
      />
      <div
        class="taro-img corn-img corn-checkbox-icon"
      >
        <img
          class="taro-img__mode-scaletofill"
          src="https://assets.xrkmm.cn/u/2000001252935290/dc949eed-eb8c-4ea5-bea4-c45a870cf0e7.png"
        />
      </div>
      选项2
    </div>
  </div>
</div>
`;

exports[`基本使用 1`] = `
<div>
  <div
    class=""
  >
    <input
      class="corn-checkbox-group-hide"
      data-testid="check"
      readonly=""
      value="[]"
    />
    <div
      class="corn-if corn-ac corn-checkbox"
      data-testid="corn-checkbox"
    >
      <input
        class="corn-checkbox-hide"
        data-testid="check"
        name=""
        readonly=""
        value="1"
      />
      <div
        class="taro-img corn-img corn-checkbox-icon"
      >
        <img
          class="taro-img__mode-scaletofill"
          src="https://assets.xrkmm.cn/u/2000001252935290/dc949eed-eb8c-4ea5-bea4-c45a870cf0e7.png"
        />
      </div>
      选项1
    </div>
    <div
      class="corn-if corn-ac corn-checkbox"
      data-testid="corn-checkbox"
    >
      <input
        class="corn-checkbox-hide"
        data-testid="check"
        name=""
        readonly=""
        value="2"
      />
      <div
        class="taro-img corn-img corn-checkbox-icon"
      >
        <img
          class="taro-img__mode-scaletofill"
          src="https://assets.xrkmm.cn/u/2000001252935290/dc949eed-eb8c-4ea5-bea4-c45a870cf0e7.png"
        />
      </div>
      选项2
    </div>
  </div>
</div>
`;

exports[`选中 1`] = `
<div>
  <div
    class=""
  >
    <input
      class="corn-checkbox-group-hide"
      data-testid="check"
      readonly=""
      value="[\\"1\\",\\"2\\"]"
    />
    <div
      class="corn-if corn-ac corn-checkbox"
      data-testid="corn-checkbox"
    >
      <input
        class="corn-checkbox-hide"
        data-testid="check"
        name=""
        readonly=""
        value="1"
      />
      <div
        class="taro-img corn-img corn-checkbox-icon"
      >
        <img
          class="taro-img__mode-scaletofill"
          src="https://assets.xrkmm.cn/u/2000001252935290/97f184d7-4fda-43f1-bd56-8313ee8a2b94.png"
        />
      </div>
      选项1
    </div>
    <div
      class="corn-if corn-ac corn-checkbox"
      data-testid="corn-checkbox"
    >
      <input
        class="corn-checkbox-hide"
        data-testid="check"
        name=""
        readonly=""
        value="2"
      />
      <div
        class="taro-img corn-img corn-checkbox-icon"
      >
        <img
          class="taro-img__mode-scaletofill"
          src="https://assets.xrkmm.cn/u/2000001252935290/97f184d7-4fda-43f1-bd56-8313ee8a2b94.png"
        />
      </div>
      选项2
    </div>
  </div>
</div>
`;
